### PR TITLE
TE-11.3  implemented changes as per PR 1158

### PR DESCRIPTION
**Pull request by bkreddy143**
Implement the updated test plan per #1158 

1. Added port 4 to topology and modified vrf names
2. Added port 1 to VRF-1
2. Added Decap and Encap support
3. Added backup support primary to decap+encap and decap+encap to decap
4. Added packet validation based on path
5. Added lookup vrf support in AddNH funtion and also optimized the code This code is a Contribution to the OpenConfig 
6. Also merged README.md file from original PR#1158
Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind.

### CODE REVIEW

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go [hunk 1]
```diff
@@ -141,29 +171,54 @@ func configureDUT(t *testing.T, dut *ondatra.DUTDevice) {
...
+		fptest.AssignToNetworkInstance(t, dut, p4.Name(), *deviations.DefaultNetworkInstance, 0)
+	}
+
 }
 
 // addStaticRoute configures static route.
 func addStaticRoute(t *testing.T, dut *ondatra.DUTDevice) {
 	d := gnmi.OC()
 	s := &oc.Root{}
 	static := s.GetOrCreateNetworkInstance(*deviations.DefaultNetworkInstance).GetOrCreateProtocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_STATIC, *deviations.StaticProtocolName)
```
Comment (cprabha) [inline feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:195]: Default network instance should have config type as "DEFAULT_INSTANCE".

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go [hunk 2]
```diff
@@ -89,6 +108,34 @@ var (
...
+		"port1": atePort1,
+		"port2": atePort2,
+		"port3": atePort3,
+		"port4": atePort4,
+	}
+	dutPorts = map[string]attrs.Attributes{
+		"port1": dutPort1,
+		"port2": dutPort2,
+		"port3": dutPort3,
+		"port4": dutPort4,
```
[the remark this answers is not in the document]
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:137]: Read me will get updated as part of PR-1158 once that is merged.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go [hunk 3]
```diff
@@ -18,18 +18,37 @@ import (
...
+	mask            = "32"
+	outerDstIP1     = "198.51.100.1"
+	outerSrcIP1     = "199.51.100.1"
+	outerDstIP2     = "111.111.111.111"
+	outerSrcIP2     = "222.222.222.222"
+	innerDstIP1     = "98.51.100.1"
+	innerSrcIP1     = "99.51.100.1"
+	VIP1            = "100.1.1.1"
+	VIP2            = "101.1.1.1"
+	vrf1            = "VRF-A"
```
Comment (xw-g) [inline feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:32]: Shall we make the variable names as `vrfA`, `vrfB` and `vrfC` for the easy of reading?

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go [hunk 4]
```diff
@@ -257,30 +320,53 @@ func testBackupDecap(ctx context.Context, t *testing.T, args *testArgs) {
 	if want := oc.Interface_OperStatus_DOWN; operStatus != want {
 		t.Errorf("Get(DUT port2 oper status): got %v, want %v", operStatus, want)
 	}
-
-	t.Log("Validate traffic passes through port3")
-	validateTrafficFlows(t, args.ate, backupFlow, baselineFlow)
+	t.Run("ValidateDecapPath", func(t *testing.T) {
+		t.Log("Validate traffic passes through ATE port4 after decap")
```
Comment (xw-g) [inline feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:324]: Same comment about log message as above.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go [hunk 5]
```diff
@@ -257,30 +320,53 @@ func testBackupDecap(ctx context.Context, t *testing.T, args *testArgs) {
...
+	args.client.AddNH(t, NH1ID, atePort2.IPv4, *deviations.DefaultNetworkInstance, fluent.InstalledInFIB)
+	args.client.AddNHG(t, NHG1ID, map[uint64]uint64{NH1ID: 1}, *deviations.DefaultNetworkInstance, fluent.InstalledInFIB)
+	args.client.AddIPv4(t, VIP1+"/"+mask, NHG1ID, *deviations.DefaultNetworkInstance, *deviations.DefaultNetworkInstance, fluent.InstalledInFIB)
+
+	t.Logf("Adding VIP %v/32 with NHG %d , NH %d and  atePort3 via gRIBI", VIP2, NHG2ID, NH2ID)
+	args.client.AddNH(t, NH2ID, atePort3.IPv4, *deviations.DefaultNetworkInstance, fluent.InstalledInFIB)
+	args.client.AddNHG(t, NHG2ID, map[uint64]uint64{NH2ID: 1}, *deviations.DefaultNetworkInstance, fluent.InstalledInFIB)
+	args.client.AddIPv4(t, VIP2+"/"+mask, NHG2ID, *deviations.DefaultNetworkInstance, *deviations.DefaultNetworkInstance, fluent.InstalledInFIB)
+
+	t.Logf("Adding NHG %d with NH %d as decap and DEFAULT vrf lookup via gRIBI", NHG100ID, NH100ID)
```
Comment (xw-g) [inline feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:342]: Typo in the log message? NHG100 does not do decap.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go (hunk 1 above)
Comment (LimeHat) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:195]: @cprabha this is not related to the lines your highlighted. 
Changes that were introduced as a part of #1335 will be added during branch merge with the upstream main; this pr is branched off #1158 and will be merged into it.

**Review by xw-g — COMMENTED** [review]
If it makes things convenient for you, I'm ok with copy&paste PR#1158 README here and close PR#1158, so that you can directly pull into main.  (But you need to mention PR#1158 in your PR description). 

Also, need to update OTG test. Thanks.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cprabha) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:137]: @xw-g , check PR # 1158 readme looks good.

**Review by greg-dennis — CHANGES_REQUESTED** [review]

**Comment by cprabha** [comment]
> @bkreddy143 thanks for the contribution. This is good. Overall looks good to me, just some minor comment.
> 
> @cprabha can you help review too? Any other concerns? (non-blocking or vendor specific changes can come after later with incremental PRs)

@xw-g , sure, will check it. Thank you

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go [hunk 6]
```diff
@@ -257,30 +320,53 @@ func testBackupDecap(ctx context.Context, t *testing.T, args *testArgs) {
...
-	t.Logf("Adding an IPv4Entry for %s with primary DecapEncap & backup decap via gRIBI", primaryTunnelDstIP)
-	args.client.AddIPv4(t, primaryTunnelDstIP+"/"+mask, NH1ID, vrfName, *deviations.DefaultNetworkInstance, fluent.InstalledInFIB)
+	t.Logf("Adding NHG %d NH %d and  NH as decap and encap with destination vrf as %v and backup NHG %d via gRIBI", NHG102ID, NH102ID, vrf3, NHG103ID)
+	args.client.AddNH(t, NH102ID, "DecapEncap", *deviations.DefaultNetworkInstance, fluent.InstalledInFIB, &gribi.NHOptions{Src: outerSrcIP2, Dest: outerDstIP2, VrfName: vrf3})
+	args.client.AddNHG(t, NHG102ID, map[uint64]uint64{NH102ID: 1}, *deviations.DefaultNetworkInstance, fluent.InstalledInFIB, &gribi.NHGOptions{BackupNHG: NHG103ID})
+
+	t.Logf("Adding an IPv4Entry for %s in vrf %s with decap and encap destiantion  in  %s via gRIBI", outerDstIP1, vrf2, vrf3)
+	args.client.AddIPv4(t, outerDstIP1+"/"+mask, NHG102ID, vrf2, *deviations.DefaultNetworkInstance, fluent.InstalledInFIB)
+
+	t.Logf("Adding NHG %d NH %d and  NH as %v  and backup NHG %d via gRIBI", NHG104ID, NH104ID, VIP2, NHG103ID)
```
Comment (cprabha) [inline feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:364]: Could you please move the programming of NH104, NHG104 and outerDstIP2 (L#364 - L#369) before programming NH102 (L#356) ? This is to ensure "Dest: outerDstIP2" is programmed before using it.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go [hunk 7]
```diff
@@ -18,18 +18,37 @@ import (
...
-	NH1ID                = 1
-	NH2ID                = 2
-	NH3ID                = 3
-	innersrcPfx          = "198.51.100.1"
+	ipv4PrefixLen   = 30
+	ipv6PrefixLen   = 126
+	mask            = "32"
+	outerDstIP1     = "198.51.100.1"
+	outerSrcIP1     = "199.51.100.1"
+	outerDstIP2     = "111.111.111.111"
```
[the remark this answers is not in the document]
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: Updated IPs as per guidelines.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go [hunk 8]
```diff
@@ -18,18 +18,37 @@ import (
...
+	outerSrcIP2     = "222.222.222.222"
+	innerDstIP1     = "98.51.100.1"
+	innerSrcIP1     = "99.51.100.1"
+	VIP1            = "100.1.1.1"
+	VIP2            = "101.1.1.1"
+	vrf1            = "VRF-A"
+	vrf2            = "VRF-B"
+	vrf3            = "VRF-C"
+	NH1ID           = 1
```
[the remark this answers is not in the document]
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: Update with small case variable names

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go [hunk 9]
```diff
@@ -141,29 +171,54 @@ func configureDUT(t *testing.T, dut *ondatra.DUTDevice) {
 	p3 := dut.Port(t, "port3")
 	gnmi.Replace(t, dut, d.Interface(p3.Name()).Config(), dutPort3.NewOCInterface(p3.Name()))
 
+	p4 := dut.Port(t, "port4")
```
[the remark this answers is not in the document]
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: Updated

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go (hunk 3 above)
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: updated

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go [hunk 10]
```diff
@@ -193,7 +248,7 @@ func TestBackupNHGAction(t *testing.T) {
 		},
 		{
 			name: "testDecapEncap",
-			desc: "Usecase3 with 2 NHOP Groups - Primary DecapEncap, Backup Pointing to Decap",
+			desc: "Usecase3 with 3 NHOP Groups - Redirect pointing to back up DecapEncap and its Backup Pointing to Decap",
```
[the remark this answers is not in the document]
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: corrected

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go [hunk 11]
```diff
@@ -230,21 +285,29 @@ func TestBackupNHGAction(t *testing.T) {
...
+	args.client.AddNHG(t, NHG101ID, map[uint64]uint64{NH101ID: 1}, *deviations.DefaultNetworkInstance, fluent.InstalledInFIB, &gribi.NHGOptions{BackupNHG: NHG100ID})
+
+	t.Logf("Adding an IPv4Entry for %s in VRF %s with primary atePort2, backup as Decap via gRIBI", outerDstIP1, vrf1)
+	args.client.AddIPv4(t, outerDstIP1+"/"+mask, NHG101ID, vrf1, *deviations.DefaultNetworkInstance, fluent.InstalledInFIB)
+
+	t.Logf("Create flows with dst %s", outerDstIP1)
+	baseFlow := createFlow(t, args.ate, args.top, "BaseFlow", &atePort2)
+	decapFlow := createFlow(t, args.ate, args.top, "DecapFlow", &atePort4)
+	t.Run("ValidatePrimaryPath", func(t *testing.T) {
+		t.Log("Validate primary path traffic passes through Port2")
```
[the remark this answers is not in the document]
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: updated

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go (hunk 4 above)
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: updated

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go (hunk 5 above)
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: corrected

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go [hunk 12]
```diff
@@ -290,11 +376,15 @@ func testDecapEncap(ctx context.Context, t *testing.T, args *testArgs) {
...
-	backupFlow := createFlow(t, args.ate, args.top, "BackupFlow", primaryTunnelDstIP, &atePort3)
-	t.Logf("Validate traffic passes through port2")
-	validateTrafficFlows(t, args.ate, baselineFlow, backupFlow)
+	t.Logf("Create flows with dst %s", outerDstIP1)
+	baseFlow := createFlow(t, args.ate, args.top, "BaseFlow", &atePort2)
+	encapFLow := createFlow(t, args.ate, args.top, "DecapEncapFlow", &atePort3)
+	decapFLow := createFlow(t, args.ate, args.top, "DecapFlow", &atePort4)
+
+	t.Run("ValidatePrimaryPath", func(t *testing.T) {
+		t.Logf("Validate Primary path traffic passes through port2")
```
[the remark this answers is not in the document]
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: updated

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go (hunk 6 above)
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go]: Corrected

File: internal/gribi/gribi.go [hunk 13]
```diff
@@ -201,12 +201,24 @@ func (c *Client) AddNHG(t testing.TB, nhgIndex uint64, nhWeights map[uint64]uint
 func (c *Client) AddNH(t testing.TB, nhIndex uint64, address, instance string, expectedResult fluent.ProgrammingResult, opts ...*NHOptions) {
 	t.Helper()
 	switch address {
+	case "Vrf":
+		NH := fluent.NextHopEntry().
```
[the remark this answers is not in the document]
Comment (bkreddy143) [reply internal/gribi/gribi.go]: @greg-dennis  are you taking about PR-1181 ?  In this PR i have optimized the code and also added validation for each case where as existing code validate modify response only for default case .  Also update variable to smaller case.

File: feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bkreddy143) [reply feature/experimental/gribi/ate_tests/backup_nhg_action/backup_nhg_action_test.go:137]: RAEDME.md  changes from PR-1158 are merged to this PR and will be closing 1158

**Comment by xw-g** [comment]
@cprabha approve?

File: internal/gribi/gribi.go [hunk 14]
```diff
@@ -200,42 +200,41 @@ func (c *Client) AddNHG(t testing.TB, nhgIndex uint64, nhWeights map[uint64]uint
 // AddNH adds a NextHopEntry with a given index to an address within a given network instance.
 func (c *Client) AddNH(t testing.TB, nhIndex uint64, address, instance string, expectedResult fluent.ProgrammingResult, opts ...*NHOptions) {
 	t.Helper()
+	nh := fluent.NextHopEntry().
+		WithNetworkInstance(instance).
+		WithIndex(nhIndex)
 	switch address {
+	case "Vrf":
```
Comment (LimeHat) [inline internal/gribi/gribi.go:207]: Let's use `VRFOnly` to match the updated te-11.1 and te-11.2 tests (seems all three are about to be merged)

**Review by cprabha — APPROVED** [review]
Looks good.

**Comment by xw-g** [comment]
@mojiiba ok to merge? If any deviation needed or vendor specific, let's use other PRs to iterate. Thanks.

**Comment by vbhan-cisco** [comment]
You can merge this PR. Moji is away today

**Comment by xw-g** [comment]
@vosipchu also just gave LGTM. 

Given that this is a change of test plan and we have many reviewers already agreed with the logic and code changes, I'll go ahead merge this PR.

**Review by xw-g — DISMISSED** [review]

**Comment by xw-g** [comment]
@bkreddy143 can you help address the merge conflict? Thanks.

**Review by greg-dennis — APPROVED** [review]

File: internal/gribi/gribi.go (hunk 14 above)
Comment (bkreddy143) [reply internal/gribi/gribi.go]: Updated

**Comment by bkreddy143** [comment]
> @bkreddy143 can you help address the merge conflict? Thanks.
> 
> I think the merge conflict probably is because this gribi.go file: https://github.com/openconfig/featureprofiles/pull/1181/files#diff-e495bef600e4cd54f409c7546966272dd971aed71b6d649251d3eb16cf2c4b78.
> 
> Let me know if this is not the case, and the conflict actually changed the test case and requires another review. Thanks.


@xw-g  , I resolved conflicts .

**Review by xw-g — APPROVED** [review]

**Pull request merged**